alkemio_virtual_contributor_engine/rabbitmq.py

```diff
@@ -54,16 +57,50 @@ async def consume(self, callback):
 
     async def publish(self, message):
         try:
-            await self.exchange.publish(
-                Message(
-                    body=json.dumps(message).encode(),
-                ),
-                routing_key=env.rabbitmq_result_routing_key,
-            )
+            if self.channel is not None and self.channel.is_closed:
+                await self.channel.reopen()
+            if self.exchange is not None:
+                await self.exchange.publish(
+                    Message(
+                        body=json.dumps(message).encode(),
+                    ),
+                    routing_key=env.rabbitmq_result_routing_key,
+                )
         except (
             aio_pika.exceptions.AMQPError,
             asyncio.exceptions.CancelledError,
             aiormq.exceptions.ChannelInvalidStateError,
             Exception,
         ) as e:
+            logger.error(e)
```

Comment:
Do you need both logged?

Reply:
So `logger.error(e)` gives us a good stack trace. The string message tells us what operation has gone wrong. Putting the error in the string message as f-string doesn't work as the stack trace is unusable. I've unified the logs in the whole client -
- log the exception
- log a message about what's wrong